File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/signing/core/SimDeltas.kt
package app.getvela.wallet.feature.signing.core

import app.getvela.wallet.feature.wallet.core.TrustAssetDelta
import app.getvela.wallet.feature.wallet.core.TrustDeltaKind
import java.math.BigInteger
import org.json.JSONArray
import org.json.JSONObject

/**
 * The simulation's shell half (spec 046 D1, the desktop's `executor/sim.rs`
 * verbatim): the `eth_simulateV1` payload for the request's calls, and the
 * net asset deltas from the logs of the calls that succeeded. Transfer logs
 * touching the wallet are netted per token; the `0xeeee…` sender is the
 * native coin. What the deltas MEAN — trusted, unverified, how to show a
 * received amount — is the core's (`token_trust::judge_delta`); nothing
 * here is ever written anywhere.
 */
object SimDeltas {
    const val TRANSFER_TOPIC = "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef"
    const val NATIVE_SENTINEL = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee"

    class Call(val to: String, val value: String?, val data: String?)

    /** The `eth_simulateV1` params: one block-state call carrying every call; `null` when there is nothing to simulate. */
    fun payload(from: String, calls: List<Call>): JSONArray? {
        val first = calls.firstOrNull() ?: return null
        if (first.to.isEmpty()) return null
        val entries = JSONArray()
        calls.forEach { call ->
            val entry = JSONObject().put("from", from).put("to", call.to).put("value", hexValue(call.value))
            val data = call.data.orEmpty()
            if (data.isNotEmpty() && data != "0x") entry.put("data", data)
            entries.put(entry)
        }
        val block = JSONObject().put("calls", entries)
        val body = JSONObject()
            .put("blockStateCalls", JSONArray().put(block))
            .put("validation", false)
            .put("traceTransfers", true)
            .put("returnFullTransactions", false)
        return JSONArray().put(body).put("latest")
    }

    /** The logs of every succeeded call in every simulated block; `null` when the answer is an error or empty. */
    fun logsOf(result: JSONObject): List<JSONObject>? {
        if (result.opt("error") != null && !result.isNull("error")) return null
        val blocks = result.optJSONArray("result") ?: return null
        if (blocks.length() == 0) return null
        val logs = ArrayList<JSONObject>()
        for (b in 0 until blocks.length()) {
            val calls = blocks.optJSONObject(b)?.optJSONArray("calls") ?: continue
            for (c in 0 until calls.length()) {
                val call = calls.optJSONObject(c) ?: continue
                if (!succeeded(call)) continue
                val entries = call.optJSONArray("logs") ?: continue
                for (l in 0 until entries.length()) entries.optJSONObject(l)?.let(logs::add)
            }
        }
        return logs
    }

    private fun succeeded(call: JSONObject): Boolean = when (val status = call.opt("status")) {
        is String -> status == "0x1"
        is Number -> status.toLong() == 1L
        else -> call.opt("error") == null || call.isNull("error")
    }

    /** Net Transfer deltas for `user`, in first-seen order, zero totals dropped. */
    fun deriveDeltas(logs: List<JSONObject>, user: String): List<TrustAssetDelta> {
        val me = user.lowercase()
        val order = ArrayList<String>()
        val totals = HashMap<String, BigInteger>()
        for (log in logs) {
            val topics = log.optJSONArray("topics") ?: continue
            if (topics.length() != 3) continue
            if (!topics.optString(0).equals(TRANSFER_TOPIC, ignoreCase = true)) continue
            val from = topicAddress(topics.optString(1))
            val to = topicAddress(topics.optString(2))
            if (from != me && to != me) continue
            val value = firstWord(log.optString("data")) ?: continue
            if (value.signum() == 0) continue
            val address = log.optString("address").lowercase()
            if (address.isEmpty()) continue
            val key = if (address == NATIVE_SENTINEL) "native" else address
            if (key !in totals) order.add(key)
            var total = totals[key] ?: BigInteger.ZERO
            if (to == me) total += value
            if (from == me) total -= value
            totals[key] = total
        }
        return order.mapNotNull { key ->
            val delta = totals[key] ?: BigInteger.ZERO
            if (delta.signum() == 0) return@mapNotNull null
            if (key == "native") {
                TrustAssetDelta(kind = TrustDeltaKind.Native, token = null, delta = delta.toString())
            } else {
                TrustAssetDelta(kind = TrustDeltaKind.Erc20, token = key, delta = delta.toString())
            }
        }
    }

    /** A 32-byte topic → the address in its low 20 bytes, lowercase. */
    internal fun topicAddress(topic: String): String {
        val hex = topic.removePrefix("0x").lowercase()
        return if (hex.length >= 40) "0x" + hex.takeLast(40) else ""
    }

    /** The first 32-byte word of `data` as an unsigned integer. */
    internal fun firstWord(data: String?): BigInteger? {
        val hex = data.orEmpty().removePrefix("0x")
        if (hex.isEmpty()) return null
        val word = hex.take(64)
        return runCatching { BigInteger(word, 16) }.getOrNull()
    }

    /** A value as the node wants it: `0x`-hex; decimal input converted; empty → `0x0`. */
    internal fun hexValue(value: String?): String {
        val raw = value.orEmpty().trim()
        if (raw.isEmpty()) return "0x0"
        if (raw.startsWith("0x") || raw.startsWith("0X")) {
            val digits = raw.drop(2).trimStart('0')
            return "0x" + digits.ifEmpty { "0" }
        }
        val parsed = raw.toBigIntegerOrNull() ?: return "0x0"
        return "0x" + parsed.toString(16)
    }
}

File: app-ios/VelaWallet/VelaWallet/Features/Signing/Core/SimDeltas.swift
//
//  SimDeltas.swift
//  VelaWallet
//
//  The transaction simulation's SHELL half (spec 055 D2).
//
//  Ported from `app-desktop/.../executor/sim.rs` and
//  `app-android/.../feature/signing/core/SimDeltas.kt`. Two jobs and no third:
//  build the `eth_simulateV1` payload for a request's calls, and net the
//  `Transfer` logs of the calls that succeeded into per-token deltas.
//
//  **What the deltas MEAN is the core's** (`token_trust::judge_delta`), and its
//  judgment is asymmetric on purpose: an outflow renders whenever the token's
//  metadata resolved, because the real token emits its own log and an outflow
//  cannot be understated; an INFLOW renders a confident number only for a token
//  already in the trusted set, because a log is something anybody can emit.
//
//  **Nothing here is ever written anywhere.** A simulated delta may not add a
//  token to somebody's list — that entrance belongs to the confirmed receipt
//  alone (spec 017, invariant ⑤).
//

import Foundation

enum SimDeltas {

    /// `keccak256("Transfer(address,address,uint256)")`.
    static let transferTopic = "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef"

    /// The sender a node reports for a NATIVE value move under
    /// `traceTransfers` — not a contract, a sentinel.
    static let nativeSentinel = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee"

    /// One leg of the operation being simulated.
    struct Call {
        let to: String
        let value: String?
        let data: String?
    }

    /// The `eth_simulateV1` params: ONE block-state call carrying every leg, so
    /// the legs see each other's effects — which is the whole reason to
    /// simulate a batch rather than each call alone.
    ///
    /// `nil` when there is nothing to simulate.
    static func payload(from: String, calls: [Call]) -> [Any]? {
        guard let body = body(from: from, calls: calls) else { return nil }
        return [body, "latest"]
    }

    static func body(from: String, calls: [Call]) -> [String: Any]? {
        guard let first = calls.first, !first.to.isEmpty else { return nil }
        let entries: [[String: Any]] = calls.map { call in
            var entry: [String: Any] = [
                "from": from, "to": call.to, "value": hexValue(call.value),
            ]
            let data = call.data ?? ""
            if !data.isEmpty, data != "0x" { entry["data"] = data }
            return entry
        }
        return [
            "blockStateCalls": [["calls": entries]],
            // No nonce or balance checks: the question is what the calls DO,
            // not whether this account could pay for them right now.
            "validation": false,
            // Native moves become logs, which is what makes them countable.
            "traceTransfers": true,
            "returnFullTransactions": false,
        ]
    }

    /// The logs of every SUCCEEDED call in every simulated block.
    ///
    /// `nil` when the node answered an error or nothing at all — which the
    /// sheet must show as "could not check", never as "nothing moves".
    static func logsOf(_ result: [String: Any]) -> [[String: Any]]? {
        if let error = result["error"], !(error is NSNull) { return nil }
        guard let blocks = result["result"] as? [[String: Any]], !blocks.isEmpty else { return nil }
        var logs: [[String: Any]] = []
        for block in blocks {
            guard let calls = block["calls"] as? [[String: Any]] else { continue }
            for call in calls where succeeded(call) {
                logs += (call["logs"] as? [[String: Any]]) ?? []
            }
        }
        return logs
    }

    /// A call is a success unless the node says otherwise. Two spellings of the
    /// status are in the wild (`"0x1"` and a number), and a call with neither
    /// and no error is one that ran.
    private static func succeeded(_ call: [String: Any]) -> Bool {
        if let status = call["status"] as? String { return status == "0x1" }
        if let status = call["status"] as? NSNumber { return status.intValue == 1 }
        let error = call["error"]
        return error == nil || error is NSNull
    }

    /// Net `Transfer` deltas for one address, in first-seen order, zeros
    /// dropped.
    ///
    /// In-and-out of the same token nets to nothing, and printing a pair of
    /// moves that cancel is how a swap looks like a theft.
    static func deriveDeltas(logs: [[String: Any]], user: String) -> [[String: Any]] {
        let me = user.lowercased()
        var order: [String] = []
        var totals: [String: SignedDigits] = [:]

        for log in logs {
            guard let topics = log["topics"] as? [String], topics.count == 3,
                  topics[0].caseInsensitiveCompare(transferTopic) == .orderedSame
            else { continue }
            let from = topicAddress(topics[1])
            let to = topicAddress(topics[2])
            guard from == me || to == me else { continue }
            guard let value = SignedDigits.firstWord(hex: log["data"] as? String ?? ""),
                  !value.isZero
            else { continue }
            let address = (log["address"] as? String ?? "").lowercased()
            guard !address.isEmpty else { continue }
            let key = address == nativeSentinel ? "native" : address
            if totals[key] == nil { order.append(key) }
            var total = totals[key] ?? .zero
            if to == me { total = total.adding(value) }
            if from == me { total = total.subtracting(value) }
            totals[key] = total
        }

        return order.compactMap { key in
            guard let total = totals[key], !total.isZero else { return nil }
            return key == "native"
                ? ["kind": "native", "token": NSNull(), "delta": total.text]
                : ["kind": "erc20", "token": key, "delta": total.text]
        }
    }

    /// A 32-byte topic → the address in its low 20 bytes, lowercase.
    static func topicAddress(_ topic: String) -> String {
        let hex = (topic.hasPrefix("0x") ? String(topic.dropFirst(2)) : topic).lowercased()
        return hex.count >= 40 ? "0x" + String(hex.suffix(40)) : ""
    }

    /// A value as the node wants it: `0x` hex, decimal input converted, empty
    /// and unreadable both `0x0` — a call with no value moves none.
    static func hexValue(_ value: String?) -> String {
        let raw = (value ?? "").trimmingCharacters(in: .whitespaces)
        guard !raw.isEmpty else { return "0x0" }
        if raw.hasPrefix("0x") || raw.hasPrefix("0X") {
            let digits = String(raw.dropFirst(2).drop { $0 == "0" })
            return "0x" + (digits.isEmpty ? "0" : digits)
        }
        guard let parsed = SignedDigits(raw), !parsed.negative else { return "0x0" }
        return "0x" + hexDigits(parsed)
    }

    /// A non-negative decimal string as lowercase hex, by repeated halving —
    /// the values are past `UInt64` and this is the only conversion needed.
    private static func hexDigits(_ value: SignedDigits) -> String {
        var digits = Array(value.digits).map { $0.wholeNumberValue ?? 0 }
        guard digits.contains(where: { $0 != 0 }) else { return "0" }
        var out = ""
        while digits.contains(where: { $0 != 0 }) {
            var remainder = 0
            var next: [Int] = []
            for digit in digits {
                let current = remainder * 10 + digit
                next.append(current / 16)
                remainder = current % 16
            }
            out.append(Character(String(remainder, radix: 16)))
            while next.first == 0, next.count > 1 { next.removeFirst() }
            digits = next
        }
        return String(out.reversed())
    }
}
